src/server/utils/logging/otel/formatter.js

```diff
@@ -0,0 +1,125 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import os from 'os';
+import util from 'util';
+import { SeverityNumber } from '@opentelemetry/api-logs';
+import flatten from 'flat';
+
+function getBaseRecord(level) {
+  let severityText = level.toUpperCase();
+  if (!Object.keys(SeverityNumber).includes(severityText)) severityText = 'INFO';
+  return {
+    attributes: {
+      agent: `${os.type()} ${os.arch()}`,
+    },
+    severityText,
+    severityNumber: SeverityNumber[severityText],
+  };
+}
+
+function generateErrorField(err) {
+  return {
+    name: err.name,
+    // doesn't accept `null`
+    message: err.message || '<none>',
+    stacktrace: err.stack || '<none>',
+    // TODO: parse the stacktrace for the file, line, and column?
+    // throwLocation: {
+    //   file: 'browser.js',
+    //   lineNumber: 1234,
+    //   columnNumber: 123,
+    //   method: 'functionName',
+    // },
+  };
+}
+
+function leadingErrorWithOtherArgs(level, ...args) {
+  const [err] = args;
+  const record = getBaseRecord(level);
+  record.body = err.message;
+  record.attributes.error = generateErrorField(err);
+
+  if (err.name === 'ClientReportedError') {
+    record.attributes.device = {
+      agent: err.userAgent,
+    };
+    record.attributes.request = {
+      address: {
+        uri: err.uri,
+      },
+      metaData: err.metaData,
+    };
+  } else if (err.metaData) {
+    record.attributes.metaData = err.metaData;
+  }
+
+  if (args.length > 1) {
+    const allButFirst = [...args];
+    allButFirst.shift();
+    record.body = util.format(...allButFirst);
+  }
+  return record;
+}
+
+function leadingStringWithError(level, ...args) {
+  const record = getBaseRecord(level);
+  record.attributes.error = generateErrorField(args[1]);
+  if (args.length === 2) {
+    [record.body] = args;
+  } else {
+    record.body = util.format(...args);
+  }
+  return record;
+}
+
+function recordFromLogType(level, logType) {
+  const record = getBaseRecord(level);
+  switch (logType.type) {
+    case 'request':
+      record.attributes.request = logType.request;
+      break;
+    default:
+      return false;
+  }
+  return record;
+}
+
+const formatter = (level, ...args) => {
+  const [arg0, arg1] = args;
+  let record;
+
+  if (arg0 instanceof Error) {
+    record = leadingErrorWithOtherArgs(level, ...args);
+  }
+  if (typeof arg0 === 'string' && arg1 instanceof Error) {
+    record = leadingStringWithError(level, ...args);
+  }
+  if (typeof arg0 === 'object' && arg0 && arg0.type) {
+    record = recordFromLogType(level, arg0);
+  }
+
+  if (!record) {
+    record = getBaseRecord(level);
+    record.body = util.format(...args);
+  }
+
+  record.attributes = flatten(record.attributes);
```

Comment:
maybe
```suggestion
  // flatten takes a deeply-nested object and returns an object of values one level deep with period-delimited keys
  record.attributes = flatten(record.attributes);
```
the example in https://www.npmjs.com/package/flat covers this so might not be needed, but FWIW my initial reaction was to look at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat and figure out if there was a new method for `Object` (I don't expect there to be in the future, this is a bit specific so seems best as a package/library 👍)